docs/src/references/agent/device-profile.md

```diff
@@ -0,0 +1,327 @@
+                "name": "collectd",
+                "version": "latest",
+                "type": "apt",
+                "action": "install"
+              }
+            ]
+          }
+        ]
+      }
+    },
+    {
+      "operation": "config_update",
+      "skip": false,
+      "payload": {
+        "type": "collectd.conf",
+        "remoteUrl":"https://abc.com/some/collectd/conf",
+        "path": "/etc/collectd/collectd.conf"
+      }
+    },
+    {
+      "operation": "software_update",
+      "skip": false,
+      "payload": {
+        "updateList": [
+          {
+            "type": "apt",
+            "modules": [
+              {
+                "name": "jq",
+                "version": "latest",
+                "action": "install"
+              }
+            ]
+          }
+        ]
+      }
+    }
+  ]
+}'
+```
+
+The profile definition in the payload is an array of operations.
+Each operation could be a `firmware_update`, `software_update` or `config_update`.
+The operations are executed in the order in which they are defined in the profile definition, by default.
+There is no restriction on the order of modules in a profile and hence can be defined in any preferred order.
+For example, additional software can be installed or configurations updated before the firmware is updated.
+This default execution order can also be overridden in the workflow definition, by updating the order in the `scheduled` state.
+
+The `"skip"` field is optional and the value is `false`, by default.
+It can be used to skip any operations during the development/testing phase, without fully deleting the entry from the profile.
+
+## Tedge agent handling device profile commands
+
+The `tedge-agent` handles `device_profile` commands using the workflow definition at `/etc/tedge/operations/device_profile.toml`.
+This workflow definition handles each module type using sub-operation workflows defined for that type.
+For example, the firmware module is installed by triggering a `firmware_update` sub-command
+which in turn uses the `firmware_update` workflow for that operation execution.
+Similarly software modules are installed with `software_update` subcommands and
+configuration updates are applied using `config_update` subcommands.
+These subcommands are triggered for each module defined in the profile definition in that order.
+
+Here is a sample device profile workflow:
+
+```toml
+operation = "device_profile"
+
+[init]
+action = "proceed"
+on_success = "scheduled"
+
+# Sort the inputs as desired
+[scheduled]
+script = "/etc/tedge/operations/device_profile.sh ${.payload.status} ${.payload}"
+on_success = "executing"
+on_error = { status = "failed", reason = "fail to sort the profile list"}
+
+[executing]
+action = "builtin"
+on_success = "next_operation"
+
+[next_operation]
+action = "iterator"
+on_next = "apply_operation"
+on_success = "successful"
+on_error = { status = "failed", reason = "Failed to compute the next operation to be executed" }
```

Comment:
The current proposal is far better but still a bit too adhoc.

- The iterator action implicitly works on a list named `operations` yielding `operation` in turn. Why not iterating over any list yielding the items of this list in turn. For example `action = "iterate @next_operation ${.operations}"` where the user can choose for the list to iterate over as well as the current item name.
- Introducing a specific `on_next` handler makes impossible to substitute a script for iterate action.

Reply:
> The iterator action implicitly works on a list named operations yielding operation in turn. Why not iterating over any list yielding the items of this list in turn. For example action = "iterate @next_operation ${.operations}" where the user can choose for the list to iterate over as well as the current item name.

Yeah, I can make this a generic iterator that takes any array as `input`, with the `@next` value generated on each iteration.

> Introducing a specific on_next handler makes impossible to substitute a script for iterate action.

I didn't fully get that point. If a user overrides this action with a script, then they don't need to use the `on_next` handle at all, but can use any number of target states using either the `on_stdout` directive or `on_exit.n` directives, right? So, I didn't really understand how this special `on_next` handle exclusive to the `iterator` action can hinder any overriding possibilities.

Reply:
Though the idea of reusing the `on_exit.n` properties is to also reduce the API, so we're not adding yet another property (which is only used in a very specific use-case).